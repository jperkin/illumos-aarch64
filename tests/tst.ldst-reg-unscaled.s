/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.12 Load/store register (unscaled immediate)
 *
 * STTRB  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRB  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRSB <Rt>, [<Xn|SP>{, #<simm>}]
 * STTRH  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRH  <Wt>, [<Xn|SP>{, #<simm>}]
 * LDTRSH <Rt>, [<Xn|SP>{, #<simm>}]
 * STTR   <Rt>, [<Xn|SP>{, #<simm>}]
 * LDTR   <Rt>, [<Xn|SP>{, #<simm>}]
 * LDTRSW <Xt>, [<Xn|SP>{, #<simm>}]
 */
.macro ldst_reg_unpriv op, wreg=0, xreg=0 simd=0
	.if \wreg
	\op	w0,  [x0]
	\op	wzr, [sp,  #-1]
	\op	w30, [x30, #-256]
	\op	wzr, [sp,  #255]
	.endif
	.if \xreg
	\op	x0,  [x0]
	\op	xzr, [sp,  #-1]
	\op	x30, [x30, #-256]
	\op	xzr, [sp,  #255]
	.endif
	.if \simd
	\op	b0,  [x0]
	\op	h31, [sp,  #0]
	\op	s0,  [x0,  #-1]
	\op	d31, [sp,  #-256]
	\op	q0,  [x0,  #255]
	.endif
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldst_reg_unpriv sturb  1
	ldst_reg_unpriv ldurb  1
	ldst_reg_unpriv ldursb 1 1
	ldst_reg_unpriv stur   1 1 1
	ldst_reg_unpriv ldur   1 1 1
	ldst_reg_unpriv sturh  1
	ldst_reg_unpriv ldurh  1
	ldst_reg_unpriv ldursh 1 1
	ldst_reg_unpriv stur   1
	ldst_reg_unpriv ldur   1
	ldst_reg_unpriv ldursw 0 1
.size libdis_test, [.-libdis_test]
