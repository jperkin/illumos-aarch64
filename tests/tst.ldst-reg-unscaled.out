    libdis_test:       00 00 00 38  sturb w0, [x0]
    libdis_test+0x4:   ff f3 1f 38  sturb wzr, [sp, #-1]
    libdis_test+0x8:   de 03 10 38  sturb w30, [x30, #-256]
    libdis_test+0xc:   ff f3 0f 38  sturb wzr, [sp, #255]
    libdis_test+0x10:  00 00 40 38  ldurb w0, [x0]
    libdis_test+0x14:  ff f3 5f 38  ldurb wzr, [sp, #-1]
    libdis_test+0x18:  de 03 50 38  ldurb w30, [x30, #-256]
    libdis_test+0x1c:  ff f3 4f 38  ldurb wzr, [sp, #255]
    libdis_test+0x20:  00 00 c0 38  ldursb w0, [x0]
    libdis_test+0x24:  ff f3 df 38  ldursb wzr, [sp, #-1]
    libdis_test+0x28:  de 03 d0 38  ldursb w30, [x30, #-256]
    libdis_test+0x2c:  ff f3 cf 38  ldursb wzr, [sp, #255]
    libdis_test+0x30:  00 00 80 38  ldursb x0, [x0]
    libdis_test+0x34:  ff f3 9f 38  ldursb xzr, [sp, #-1]
    libdis_test+0x38:  de 03 90 38  ldursb x30, [x30, #-256]
    libdis_test+0x3c:  ff f3 8f 38  ldursb xzr, [sp, #255]
    libdis_test+0x40:  00 00 00 b8  stur w0, [x0]
    libdis_test+0x44:  ff f3 1f b8  stur wzr, [sp, #-1]
    libdis_test+0x48:  de 03 10 b8  stur w30, [x30, #-256]
    libdis_test+0x4c:  ff f3 0f b8  stur wzr, [sp, #255]
    libdis_test+0x50:  00 00 00 f8  stur x0, [x0]
    libdis_test+0x54:  ff f3 1f f8  stur xzr, [sp, #-1]
    libdis_test+0x58:  de 03 10 f8  stur x30, [x30, #-256]
    libdis_test+0x5c:  ff f3 0f f8  stur xzr, [sp, #255]
    libdis_test+0x60:  00 00 00 3c  stur b0, [x0]
    libdis_test+0x64:  ff 03 00 7c  stur h31, [sp]
    libdis_test+0x68:  00 f0 1f bc  stur s0, [x0, #-1]
    libdis_test+0x6c:  ff 03 10 fc  stur d31, [sp, #-256]
    libdis_test+0x70:  00 f0 8f 3c  stur q0, [x0, #255]
    libdis_test+0x74:  00 00 40 b8  ldur w0, [x0]
    libdis_test+0x78:  ff f3 5f b8  ldur wzr, [sp, #-1]
    libdis_test+0x7c:  de 03 50 b8  ldur w30, [x30, #-256]
    libdis_test+0x80:  ff f3 4f b8  ldur wzr, [sp, #255]
    libdis_test+0x84:  00 00 40 f8  ldur x0, [x0]
    libdis_test+0x88:  ff f3 5f f8  ldur xzr, [sp, #-1]
    libdis_test+0x8c:  de 03 50 f8  ldur x30, [x30, #-256]
    libdis_test+0x90:  ff f3 4f f8  ldur xzr, [sp, #255]
    libdis_test+0x94:  00 00 40 3c  ldur b0, [x0]
    libdis_test+0x98:  ff 03 40 7c  ldur h31, [sp]
    libdis_test+0x9c:  00 f0 5f bc  ldur s0, [x0, #-1]
    libdis_test+0xa0:  ff 03 50 fc  ldur d31, [sp, #-256]
    libdis_test+0xa4:  00 f0 cf 3c  ldur q0, [x0, #255]
    libdis_test+0xa8:  00 00 00 78  sturh w0, [x0]
    libdis_test+0xac:  ff f3 1f 78  sturh wzr, [sp, #-1]
    libdis_test+0xb0:  de 03 10 78  sturh w30, [x30, #-256]
    libdis_test+0xb4:  ff f3 0f 78  sturh wzr, [sp, #255]
    libdis_test+0xb8:  00 00 40 78  ldurh w0, [x0]
    libdis_test+0xbc:  ff f3 5f 78  ldurh wzr, [sp, #-1]
    libdis_test+0xc0:  de 03 50 78  ldurh w30, [x30, #-256]
    libdis_test+0xc4:  ff f3 4f 78  ldurh wzr, [sp, #255]
    libdis_test+0xc8:  00 00 c0 78  ldursh w0, [x0]
    libdis_test+0xcc:  ff f3 df 78  ldursh wzr, [sp, #-1]
    libdis_test+0xd0:  de 03 d0 78  ldursh w30, [x30, #-256]
    libdis_test+0xd4:  ff f3 cf 78  ldursh wzr, [sp, #255]
    libdis_test+0xd8:  00 00 80 78  ldursh x0, [x0]
    libdis_test+0xdc:  ff f3 9f 78  ldursh xzr, [sp, #-1]
    libdis_test+0xe0:  de 03 90 78  ldursh x30, [x30, #-256]
    libdis_test+0xe4:  ff f3 8f 78  ldursh xzr, [sp, #255]
    libdis_test+0xe8:  00 00 00 b8  stur w0, [x0]
    libdis_test+0xec:  ff f3 1f b8  stur wzr, [sp, #-1]
    libdis_test+0xf0:  de 03 10 b8  stur w30, [x30, #-256]
    libdis_test+0xf4:  ff f3 0f b8  stur wzr, [sp, #255]
    libdis_test+0xf8:  00 00 40 b8  ldur w0, [x0]
    libdis_test+0xfc:  ff f3 5f b8  ldur wzr, [sp, #-1]
    libdis_test+0x100: de 03 50 b8  ldur w30, [x30, #-256]
    libdis_test+0x104: ff f3 4f b8  ldur wzr, [sp, #255]
    libdis_test+0x108: 00 00 80 b8  ldursw x0, [x0]
    libdis_test+0x10c: ff f3 9f b8  ldursw xzr, [sp, #-1]
    libdis_test+0x110: de 03 90 b8  ldursw x30, [x30, #-256]
    libdis_test+0x114: ff f3 8f b8  ldursw xzr, [sp, #255]
